File: src/main/java/net/beams/genera/blocks/BlockGeneraCrop.java
package net.beams.genera.blocks;

import net.minecraft.block.Block;
import net.minecraft.block.BlockBush;
import net.minecraft.block.IGrowable;
import net.minecraft.block.SoundType;
import net.minecraft.block.properties.IProperty;
import net.minecraft.block.properties.PropertyInteger;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Blocks;
import net.minecraft.init.Items;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.common.IPlantable;

import java.util.Random;

/**
 * Created by ben on 3/18/17.
 */
public class BlockGeneraCrop extends BlockBush implements IGrowable {

    public static final IProperty<Integer> AGE = PropertyInteger.create("age", 0, 2);

    public BlockGeneraCrop() {
        setTickRandomly(true);
        setCreativeTab(null);
        setHardness(0.0F);
        setSoundType(SoundType.GROUND);
        disableStats();
        setDefaultState(this.blockState.getBaseState().withProperty(AGE, Integer.valueOf(0)));
    }

    public void grow(IBlockState state, World worldIn, BlockPos pos) {
        int growStage = state.getValue(AGE).intValue() + 1;
        if (growStage > 2)
            growStage = 2;
        worldIn.setBlockState(pos, state.withProperty(AGE, Integer.valueOf(growStage)));
    }

    public boolean onBlockActivated(World worldIn, BlockPos pos, IBlockState state, EntityPlayer playerIn, EnumHand hand, EnumFacing facing, float hitX, float hitY, float hitZ)
    {
        if (playerIn.getHeldItem(hand).getItem() == new ItemStack(Items.DYE, 1, 15).getItem()) {
            grow(state, worldIn, pos);
            return super.onBlockActivated(worldIn, pos, state, playerIn, hand, facing, hitX, hitY, hitZ);
        }
        return false;
    }

    @Override
    protected BlockStateContainer createBlockState() {
        return new BlockStateContainer(this, AGE);
    }

    @Override
    public boolean canGrow(World worldIn, BlockPos pos, IBlockState state, boolean isClient) {
        return state.getBlock() == Blocks.FARMLAND;
    }

    @Override
    public boolean canUseBonemeal(World worldIn, Random rand, BlockPos pos, IBlockState state) {
        return state.getValue(AGE).intValue() < 2;
    }

    @Override
    public void grow(World worldIn, Random rand, BlockPos pos, IBlockState state) {
        grow(state, worldIn, pos);
    }

    @Override
    public void updateTick(World worldIn, BlockPos pos, IBlockState state, Random rand)
    {
        super.updateTick(worldIn, pos, state, rand);

        if (worldIn.getLightFromNeighbors(pos.up()) >= 1)
        {
            int i = state.getValue(AGE).intValue();

            if (i < 5)
            {
                float f = getGrowthChance(this, worldIn, pos);

                if (rand.nextInt((int)(25.0F / f) + 1) == 0)
                {
                    worldIn.setBlockState(pos, state.withProperty(AGE, Integer.valueOf(i + 1)), 2);
                }
            }
        }
    }

    protected static float getGrowthChance(Block blockIn, World worldIn, BlockPos pos)
    {
        float f = 1.0F;
        BlockPos blockpos1 = pos.down();

        for (int i = -1; i <= 1; ++i)
        {
            for (int j = -1; j <= 1; ++j)
            {
                float f1 = 0.0F;
                IBlockState iblockstate = worldIn.getBlockState(blockpos1.add(i, 0, j));

                if (iblockstate.getBlock().canSustainPlant(worldIn.getBlockState(blockpos1.add(i, 0, j)),
                        worldIn, blockpos1.add(i, 0, j), EnumFacing.UP, (IPlantable)blockIn))
                {
                    f1 = 1.0F;

                    if (iblockstate.getBlock().isFertile(worldIn, blockpos1.add(i, 0, j)))
                    {
                        f1 = 3.0F;
                    }
                }

                if (i != 0 || j != 0)
                {
                    f1 /= 4.0F;
                }

                f += f1;
            }
        }

        BlockPos blockpos2 = pos.north();
        BlockPos blockpos3 = pos.south();
        BlockPos blockpos4 = pos.west();
        BlockPos blockpos5 = pos.east();
        boolean flag = blockIn == worldIn.getBlockState(blockpos4).getBlock() || blockIn == worldIn.getBlockState(blockpos5).getBlock();
        boolean flag1 = blockIn == worldIn.getBlockState(blockpos2).getBlock() || blockIn == worldIn.getBlockState(blockpos3).getBlock();

        if (flag && flag1)
        {
            f /= 2.0F;
        }
        else
        {
            boolean flag2 = blockIn == worldIn.getBlockState(blockpos4.north()).getBlock() || blockIn == worldIn.getBlockState(blockpos5.north()).getBlock() || blockIn == worldIn.getBlockState(blockpos5.south()).getBlock() || blockIn == worldIn.getBlockState(blockpos4.south()).getBlock();

            if (flag2)
            {
                f /= 2.0F;
            }
        }

        return f;
    }

    @Override
    public int getMetaFromState(IBlockState state)
    {
        return state.getValue(AGE).intValue();
    }
}

File: src/main/kotlin/net/bms/genera/block/BlockGeneraCrop.kt
package net.bms.genera.block

import net.minecraft.block.Block
import net.minecraft.block.BlockBush
import net.minecraft.block.IGrowable
import net.minecraft.block.properties.IProperty
import net.minecraft.block.properties.PropertyInteger
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Blocks
import net.minecraft.init.Items
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraftforge.common.IPlantable
import java.util.*

/**
 * Created by ben on 3/18/17.
 */
open class BlockGeneraCrop internal constructor() : BlockBush(), IGrowable {

    init {
        tickRandomly = true
        defaultState = this.blockState.baseState.withProperty(AGE, 0)
    }

    private fun grow(state: IBlockState, worldIn: World?, pos: BlockPos?) {
        var growStage = state.getValue(AGE) + 1
        if (growStage > 2)
            growStage = 2
        worldIn!!.setBlockState(pos!!, state.withProperty(AGE, growStage))
    }

    override fun onBlockActivated(worldIn: World?, pos: BlockPos?, state: IBlockState?, playerIn: EntityPlayer?, hand: EnumHand?, facing: EnumFacing?, hitX: Float, hitY: Float, hitZ: Float): Boolean {
        if (playerIn!!.getHeldItem(hand!!).item === ItemStack(Items.DYE, 1, 15).item) {
            grow(state!!, worldIn, pos)
            playerIn!!.getHeldItem(hand!!).count -= 1
            return super.onBlockActivated(worldIn!!, pos!!, state, playerIn!!, hand!!, facing!!, hitX, hitY, hitZ)
        }
        return false
    }

    override fun createBlockState(): BlockStateContainer {
        return BlockStateContainer(this, AGE)
    }

    override fun canGrow(worldIn: World, pos: BlockPos, state: IBlockState, isClient: Boolean): Boolean {
        return state.block === Blocks.FARMLAND
    }

    override fun canUseBonemeal(worldIn: World, rand: Random, pos: BlockPos, state: IBlockState): Boolean {
        return state.getValue(AGE) < 2
    }

    override fun grow(worldIn: World, rand: Random, pos: BlockPos, state: IBlockState) {
        grow(state, worldIn, pos)
    }

    override fun updateTick(worldIn: World, pos: BlockPos, state: IBlockState, rand: Random?) {
        super.updateTick(worldIn, pos, state, rand!!)

        if (worldIn.getLightFromNeighbors(pos.up()) >= 1) {
            val i = state.getValue(AGE)

            if (i < 2) {
                val f = getGrowthChance(this, worldIn, pos)

                if (rand.nextInt((25.0f / f).toInt() + 1) == 0) {
                    worldIn.setBlockState(pos, state.withProperty(AGE, i + 1), 2)
                }
            }
        }
    }

    override fun getMetaFromState(state: IBlockState): Int {
        return state.getValue(AGE)
    }

    override fun getStateFromMeta(meta: Int): IBlockState {
        return this.defaultState.withProperty(AGE, meta)
    }

    companion object {

        internal val AGE: IProperty<Int> = PropertyInteger.create("age", 0, 2)

        private fun getGrowthChance(blockIn: Block, worldIn: World, pos: BlockPos): Float {
            var f = 1.0f
            val blockpos1 = pos.down()

            for (i in -1..1) {
                for (j in -1..1) {
                    var f1 = 0.0f
                    val iblockstate = worldIn.getBlockState(blockpos1.add(i, 0, j))

                    if (iblockstate.block.canSustainPlant(worldIn.getBlockState(blockpos1.add(i, 0, j)),
                                    worldIn, blockpos1.add(i, 0, j), EnumFacing.UP, blockIn as IPlantable)) {
                        f1 = 1.0f

                        if (iblockstate.block.isFertile(worldIn, blockpos1.add(i, 0, j))) {
                            f1 = 3.0f
                        }
                    }

                    if (i != 0 || j != 0) {
                        f1 /= 4.0f
                    }

                    f += f1
                }
            }

            val blockpos2 = pos.north()
            val blockpos3 = pos.south()
            val blockpos4 = pos.west()
            val blockpos5 = pos.east()
            val flag = blockIn === worldIn.getBlockState(blockpos4).block || blockIn === worldIn.getBlockState(blockpos5).block
            val flag1 = blockIn === worldIn.getBlockState(blockpos2).block || blockIn === worldIn.getBlockState(blockpos3).block

            if (flag && flag1) {
                f /= 2.0f
            } else {
                val flag2 = blockIn === worldIn.getBlockState(blockpos4.north()).block || blockIn === worldIn.getBlockState(blockpos5.north()).block || blockIn === worldIn.getBlockState(blockpos5.south()).block || blockIn === worldIn.getBlockState(blockpos4.south()).block

                if (flag2) {
                    f /= 2.0f
                }
            }

            return f
        }
    }
}
